### ISSUE TITLE
super mario rendering bug when running and small other

### ISSUE THREAD

**Issue by ESPboy-edu** [deltabeard/Peanut-GB#66]
greatly improved perfomance as i see in comparition with the version 16 month before :)
thanks for your great work!

but 

1. there is a rendering bug in supermario (many other games are ok and 16 month old peanutGB renders fine but slower). 
PEANUT_GB_HIGH_LCD_ACCURACY 0 or 1 - does not matter.
static picture is ok, but when running to the right it renders very strange things on the level

2. prince of persia does not starts the game. only renders initial screen with "virgin games" logo and freezes

3. 
here is a one "]" is missing at the end
```cpp
			if(gb->hram_io[IO_LY +
					(gb->hram_io[IO_LCDC] & LCDC_OBJ_SIZE ? 0 : 8) >= OY ||
					gb->hram_io[IO_LY] + 16 < OY)
```

4.
here is the array is "uint8_t" but uses negative numbers (Arduino ESP8266 compiler does not allow this), I changed to int8_t
```cpp
	const uint8_t cart_mbc[] =
	{
		0, 1, 1, 1, -1, 2, 2, -1, 0, 0, -1, 0, 0, 0, -1, 3,
		3, 3, 3, 3, -1, -1, -1, -1, -1, 5, 5, 5, 5, 5, 5, -1
	};
```

5.
here is compiler says (can not convert "void*" to  "sprite_data *")
```cpp
const struct sprite_data *sd1 = in1, *sd2 = in2;
```

this fixes the problem
```cpp
const struct sprite_data *sd1 = (sprite_data *)in1, *sd2 = (sprite_data *)in2;
```

thanks for the great GB kernel and passion to the old school :)

**Comment by deltabeard** [comment]
Thank you all for the detailed bug report.

The offending commit causing the rendering errors in Super Mario is 6310eb807d1b3a81dd0389e7b78c83efaccfa7a2.

I will look into the other problems described shortly.

**Comment by deltabeard** [comment]
Updating the SCX register is being skipped at https://github.com/kaspermeerts/supermarioland/blob/618d00ed6c330928e106719533c6e294ae5d5726/bank0.asm#L94. I haven't been able to figure out why, so I will revert the offending commit and investigate when I more time.

**Comment by deltabeard** [comment]
Regarding the other problems listed:

2. Prince of Persia is working in the latest commit 6b4a38752ef4dee6f4deafe16512f15c75a9fec0.
![Screenshot of Prince of Persia running in Peanut-SDL](https://user-images.githubusercontent.com/3747104/204128580-93309710-0d2b-47b7-ad78-c1f3ea1957d9.png). Which peanut-gb revision/commit are you using?

3. The offending commit is reverted now. I must have forgotten to test with "high LCD accuracy" switched off. I spotted this error two weeks ago and fixed it in the v2 branch at https://github.com/deltabeard/Peanut-GB/commit/7dbc700efb01be6ac47bc98027641bd415a9e4d0, but I forgot to cherry-pick the fix...

4. Fixed in 6f9be756fab412d2af797532a422f2912fda965d.

5. Fixed in e21823b63db67f958e5dab3a0e352d3df7048e2a.

**Comment by deltabeard** [comment]
The rendering bug with Super Mario is fixed in commit 9bb2d22bef76aaa986c1f440fd114fa9536ab546. The bug was caused by the overwriting of the STAT register when the LCD was changing modes (HBlank, VBlank, etc). This meant that the check for HBlank would always fail in Super Mario, and the SCX would therefore never be set to the correct value.

The missing square bracket is fixed in 067e5fc96ef2f614e0a9cd96fb0db05d9a477dbc.

![Screenshot of Peanut-SDL with working Super Mario](https://user-images.githubusercontent.com/3747104/204161856-aa156f90-e205-4d96-b416-ca78e333da7f.png)
